File: 2002PDC/Threading/Monitor/VB/MonitorSample.vb
'==========================================================================
'	File:		Monitor.cs
'
'	Summary:	Example shows use of Monitor in threads.
'					Monitor.Enter(Object)
'					Monitor.Exit(Object)
'					Monitor.Pulse(Object)
'					Monitor.Wait(Object)
'
'----------------------------------------------------------------------------
'	This file is part of the Microsoft COM+ Runtime Samples.
'
'	Copyright (C) 1998-2000 Microsoft Corporation.  All rights reserved.
'===========================================================================
imports System 
imports System.Threading 

public module MonitorSample
	Sub Main()
		dim result As Integer = 0 		'	result initialized to say there is no error
		Dim cell As ClsCell = new ClsCell( ) 

		Dim prod As CellProd = new CellProd(cell,20) 	'	use cell for storage, produce 20 items
		Dim cons As CellCons = new CellCons(cell,20) 	'	use cell for storage, consume 20 items

		Dim producer As Thread = new Thread(new ThreadStart(AddressOf prod.ThreadRun)) 
		Dim consumer As Thread = new Thread(new ThreadStart(AddressOf cons.ThreadRun)) 
		'	threads producer and consumer have been created, but not started at this point.

		try
			producer.Start( ) 
			consumer.Start( ) 

			producer.Join( ) 		'	Join with no Stop.  Run producer thread until it is done.
			consumer.Join( ) 		'	Join with no Stop.  Run consumer thread until it is done.
		'	threads producer and consumer are dead at this point.
		catch e As ThreadStateException
			Console.WriteLine(e) 			'	display text of exception
			result = 1 						'	result says there was an error
		catch e As ThreadInterruptedException
			Console.WriteLine(e) 			'	display text of exception
			result = 1 						'	result says there was an error
		end try

		Environment.ExitCode = result 		'	eventhough Main with a return type of void, this sets
	end sub										'		a return code to parent process.


public class CellProd
	Dim cell As Clscell 							'	Field to hold cell object to be used
	Dim quantity As Integer = 1 				'	Field for "How many items to produce in cell?"

	public Sub New(box As ClsCell, request As Integer)
                MyBase.New
		cell = box 							'	Pass in what cell object to be used
		quantity = request 					'	Pass in "How many items to produce in cell?"
	end sub

	public sub ThreadRun( )
                dim looper As Integer

		for looper = 1 To quantity
			cell.WriteToCell(looper) 		'	"producing"
                next
	end sub
end class '	End of public class CellProd

public class CellCons
	dim cell as Clscell 							'	Field to hold cell object to be used
	dim quantity as integer	= 1 				'	Field for "How many items to consume from cell?"

	public sub New(box as clsCell, request as integer)
                MyBase.New
		cell = box 							'	Pass in what cell object to be used
		quantity = request 					'	Pass in "How many items to consume from cell?"
	end sub

	public sub ThreadRun( )
		dim valReturned as integer
                dim looper as integer

		for looper = 1 To quantity
			valReturned=cell.ReadFromCell( ) 	'	if you wanted to "consume" the result, see valReturned
		next
	end sub
end class	'	End of public class CellCons


public class ClsCell
	dim cellContents as integer				'	Cell Contentents
	dim readerFlag As boolean = false 	'	State Flag

	public Function ReadFromCell( ) as integer
		Monitor.Enter(Me) 			'	Enter Synchronization block
		if not readerFlag then
			try
				Monitor.Wait(Me) 		'	Waits until it gets a Monitor.Pulse(this)
			catch e As SynchronizationLockException
				Console.WriteLine(e) 
			catch e As ThreadInterruptedException
				Console.WriteLine(e) 
			end try
		end if
		Console.WriteLine("Consume: {0}",cellContents) 
		readerFlag = false 			'	Reset the State Flag to say, done
		Monitor.Pulse(Me) 		'	Pulse tells Cell.WriteToCell that Cell.ReadFromCell is done.
		Monitor.Exit(Me) 			'	Exit Synchronization block
		ReadFromCell = cellContents 
	end function	'	End of method public int ReadFromCell( )
	
	public sub WriteToCell(n as integer)
		Monitor.Enter(Me) 			'	Enter Synchronization block
		if readerFlag then
			try
				Monitor.Wait(Me) 		'	Waits until it gets a Monitor.Pulse(this)
			catch e As SynchronizationLockException
				Console.WriteLine(e) 
			catch e As ThreadInterruptedException
				Console.WriteLine(e) 
			end try
		end if
		cellContents = n 
		Console.WriteLine("Produce: {0}",cellContents) 
		readerFlag = true 			'	Reset the State Flag to say, done
		Monitor.Pulse(Me) 		'	Pulse tells Cell.ReadFromCell that Cell.WriteToCell is done.
		Monitor.Exit(Me) 			'	Exit Synchronization block
	end sub	'	End of method public void WriteToCell(int value)
end class '	End of public class Cell

end module	' MonitorSample


'Output

'Produce: 1
'Consume: 1
'Produce: 2
'Consume: 2
'Produce: 3
'Consume: 3
'Produce: 4
'Consume: 4
'Produce: 5
'Consume: 5
'Produce: 6
'Consume: 6
'Produce: 7
'Consume: 7
'Produce: 8
'Consume: 8
'Produce: 9
'Consume: 9
'Produce: 10
'Consume: 10
'Produce: 11
'Consume: 11
'Produce: 12
'Consume: 12
'Produce: 13
'Consume: 13
'Produce: 14
'Consume: 14
'Produce: 15
'Consume: 15
'Produce: 16
'Consume: 16
'Produce: 17
'Consume: 17
'Produce: 18
'Consume: 18
'Produce: 19
'Consume: 19
'Produce: 20
'Consume: 20

File: 2002B2/vcsharp/tutorials/threads/monitorsample/monitorsample.cs
// MonitorSample.cs
// This example shows use of the following methods of the C# lock keyword
// and the Monitor class 
// in threads:
//      Monitor.Pulse(Object)
//      Monitor.Wait(Object)
using System;
using System.Threading;

public class MonitorSample
{
   public static void Main(String[] args)
   {
      int result = 0;   // Result initialized to say there is no error
      Cell cell = new Cell( );

      CellProd prod = new CellProd(cell, 20);  // Use cell for storage, 
                                               // produce 20 items
      CellCons cons = new CellCons(cell, 20);  // Use cell for storage, 
                                               // consume 20 items

      Thread producer = new Thread(new ThreadStart(prod.ThreadRun));
      Thread consumer = new Thread(new ThreadStart(cons.ThreadRun));
      // Threads producer and consumer have been created, 
      // but not started at this point.

      try
      {
         producer.Start( );
         consumer.Start( );

         producer.Join( );   // Join both threads with no timeout
                             // Run both until done.
         consumer.Join( );  
      // threads producer and consumer have finished at this point.
      }
      catch (ThreadStateException e)
      {
         Console.WriteLine(e);  // Display text of exception
         result = 1;            // Result says there was an error
      }
      catch (ThreadInterruptedException e)
      {
         Console.WriteLine(e);  // This exception means that the thread
                                // was interrupted during a Wait
         result = 1;            // Result says there was an error
      }
      // Even though Main returns void, this provides a return code to 
      // the parent process.
      Environment.ExitCode = result;
   }
}

public class CellProd
{
   Cell cell;         // Field to hold cell object to be used
   int quantity = 1;  // Field for how many items to produce in cell

   public CellProd(Cell box, int request)
   {
      cell = box;          // Pass in what cell object to be used
      quantity = request;  // Pass in how many items to produce in cell
   }
   public void ThreadRun( )
   {
      for(int looper=1; looper<=quantity; looper++)
         cell.WriteToCell(looper);  // "producing"
   }
}

public class CellCons
{
   Cell cell;         // Field to hold cell object to be used
   int quantity = 1;  // Field for how many items to consume from cell

   public CellCons(Cell box, int request)
   {
      cell = box;          // Pass in what cell object to be used
      quantity = request;  // Pass in how many items to consume from cell
   }
   public void ThreadRun( )
   {
      int valReturned;
      for(int looper=1; looper<=quantity; looper++)
      // Consume the result by placing it in valReturned.
         valReturned=cell.ReadFromCell( );
   }
}

public class Cell
{
   int cellContents;         // Cell contents
   bool readerFlag = false;  // State flag
   public int ReadFromCell( )
   {
      lock(this)   // Enter synchronization block
      {
         if (!readerFlag)
         {            // Wait until Cell.WriteToCell is done producing
            try
            {
               // Waits for the Monitor.Pulse in WriteToCell
               Monitor.Wait(this);
            }
            catch (SynchronizationLockException e)
            {
               Console.WriteLine(e);
            }
            catch (ThreadInterruptedException e)
            {
               Console.WriteLine(e);
            }
         }
         Console.WriteLine("Consume: {0}",cellContents);
         readerFlag = false;   // Reset the state flag to say consuming
                               // is done.
         Monitor.Pulse(this);   // Pulse tells Cell.WriteToCell that
                                // Cell.ReadFromCell is done.
      }   // Exit synchronization block
      return cellContents;
   }
   
   public void WriteToCell(int n)
   {
      lock(this)  // Enter synchronization block
      {
         if (readerFlag)
         {      // Wait until Cell.ReadFromCell is done consuming.
            try
            {
               Monitor.Wait(this);   // Wait for the Monitor.Pulse in
                                     // ReadFromCell
            }
            catch (SynchronizationLockException e)
            {
               Console.WriteLine(e);
            }
            catch (ThreadInterruptedException e)
            {
               Console.WriteLine(e);
            }
         }
         cellContents = n;
         Console.WriteLine("Produce: {0}",cellContents);
         readerFlag = true;   // Reset the state flag to say producing
                              // is done
         Monitor.Pulse(this);  // Pulse tells Cell.ReadFromCell that 
                               // Cell.WriteToCell is done.
      }   // Exit synchronization block
   }
}
